xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -0,0 +1,213 @@
+{
+    vector<string> docids;
+    string docid;
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	for (size_t j = 0; j <= sessions[i][DOCIDS].length(); ++j) {
+	    if (sessions[i][DOCIDS][j] != ',' && sessions[i][1][j] != '\0') {
+		docid += sessions[i][DOCIDS][j];
+	    } else {
+		docids.push_back(docid);
+		docid.clear();
+	    }
+	}
+    }
+    return docids;
+}
+
+static vector<int>
+get_click_list(const vector<vector<string>> &sessions)
+{
+    vector<int> clicks;
+    string clickstring, clickcount;
+    int click = 0;
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	for (size_t j = 0; j <= sessions[i][CLICKS].length(); ++j) {
+	    if (sessions[i][CLICKS][j] == '\0') {
+		clicks.push_back(click);
+	    } else if (sessions[i][CLICKS][j] != ',') {
+		// Get clickstring of the form "docid:click_count".
+		clickstring += sessions[i][CLICKS][j];
+
+		// Get click count string.
+		int delimiter_pos = clickstring.find(':');
+		clickcount = clickstring.substr(delimiter_pos + 1,
+			     clickstring.length() - 1 - delimiter_pos);
+
+		// Convert click count string to an integer.
+		stringstream ss(clickcount);
+		ss >> click;
+	    } else {
+		clickstring.clear();
+		clicks.push_back(click);
+	    }
+	}
+    }
+    return clicks;
+}
+
+vector<vector<string>>
+SimplifiedDBN::build_sessions(string logfile)
+{
+    ifstream file;
+    file.open(logfile, ios::in);
+
+    string line;
+
+    // Skip the first line.
+    getline(file, line);
+
+    vector<vector<string>> sessions;
+    vector<string> session_entry;
+
+    // Start reading file from the second line.
+    while (getline(file, line)) {
+
+	istringstream ss(line);
+
+	vector<string> row_data;
+
+	while (ss >> std::ws) {
+	    string column_element;
+	    if (ss.peek() == '"') {
+		int pos = ss.tellg();
+		ss.seekg(pos + 1);
+		char ch;
+		while (ss.get(ch)) {
+		    if (ch == '"')
+			break;
+		    column_element += ch;
+		}
+	    } else {
+		getline(ss, column_element, ',');
+	    }
+	    row_data.push_back(column_element);
+	}
+
+	string qid = row_data[0];
+	string query = row_data[1];
+	string docids = row_data[2];
+	string clicks = row_data[4];
+
+	session_entry.push_back(qid);
+	session_entry.push_back(docids);
+	session_entry.push_back(clicks);
+
+	sessions.push_back(session_entry);
+
+	session_entry.clear();
+    }
+
+    file.close();
+    return sessions;
+}
+
+void
+SimplifiedDBN::train(vector<vector<string>> &sessions)
+{
+    map<string, map<string, map<char, array<double, 2>>>> url_rel_fractions;
+
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	string qid = sessions[i][QID];
+
+	vector<string> docids = get_docid_list(sessions);
+
+	vector<int> clicks = get_click_list(sessions);
+
+	int last_clicked_pos = clicks.size() - 1;
+
+	for (size_t j = 0; j < clicks.size(); ++j)
+	    if (clicks[j] != 0)
+		last_clicked_pos = j;
+
+	vector<int> session_clicks;
+	for (int x = 0; x <= last_clicked_pos + 1; ++x)
+	    session_clicks.push_back(clicks[x]);
```

Comment:
There doesn't seem to be any need to make a partial copy of `clicks` here - the loop below could easily just loop over clicks from `0` to `last_clicked_pos` (or `last_clicked_pos + 1` if that's really correct).